File: app/src/main/java/rusha/x/recipe/createOrEdit/EditRecipeIngredientsAdapter.java
package rusha.x.recipe.createOrEdit;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.recyclerview.widget.RecyclerView;

import java.util.Collections;
import java.util.List;

import rusha.x.databinding.EditRecipeIngredientBinding;

public class EditRecipeIngredientsAdapter
        extends RecyclerView.Adapter<EditRecipeIngredientsAdapter.IngredientViewHolder> {

    private EditRecipeViewModel viewModel;

    public EditRecipeIngredientsAdapter(EditRecipeViewModel viewModel) {
        this.viewModel = viewModel;
    }

    private List<CreateOrEditRecipeIngredient> ingredientsToAdopt = Collections.emptyList();

    public void setIngredientsToAdopt(List<CreateOrEditRecipeIngredient> ingredients) {
        ingredientsToAdopt = ingredients;
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return ingredientsToAdopt.size();
    }

    @Override
    public IngredientViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        EditRecipeIngredientBinding binding = EditRecipeIngredientBinding.inflate(inflater, parent, false);
        return new IngredientViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(IngredientViewHolder holder, int position) {
        CreateOrEditRecipeIngredient ingredientOnPosition = ingredientsToAdopt.get(position);
        holder.bind(ingredientOnPosition);
    }

    class IngredientViewHolder extends RecyclerView.ViewHolder {

        private EditRecipeIngredientBinding binding;

        public IngredientViewHolder(EditRecipeIngredientBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(CreateOrEditRecipeIngredient ingredient) {
            binding.addIngredient.setText(ingredient.product.name);
            binding.ingredientCount.setText(String.valueOf(ingredient.countInRecipe));

            binding.plus.setOnClickListener((v) -> {
                viewModel.onIngredientPlusClick(ingredient);
            });

            binding.minus.setOnClickListener((v) -> {
                viewModel.onIngredientMinusClick(ingredient);
            });

            binding.ingredientCount.setOnFocusChangeListener((v, hasFocus) -> {
                if (!hasFocus) {
                    double count = Double.parseDouble(binding.ingredientCount.getText().toString());
                    viewModel.onSetIngredientCount(ingredient, count);
                }
            });
        }
    }
}

File: app/src/main/java/rusha/x/recipe/createOrEdit/EditRecipeIngredientsAdapter.kt
package rusha.x.recipe.createOrEdit

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.edit_recipe_ingredient.view.*
import rusha.x.R

class EditRecipeIngredientsAdapter(
    private val presenter: EditRecipeViewModel
) : RecyclerView.Adapter<EditRecipeIngredientsAdapter.IngredientViewHolder>() {

    var ingredientsToAdopt: List<CreateOrEditRecipe.Ingredient> = emptyList()
    override fun getItemCount(): Int {
        return ingredientsToAdopt.size
    }

    override fun onCreateViewHolder(
        parent: ViewGroup, viewType: Int
    ): IngredientViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = inflater.inflate(
            R.layout.edit_recipe_ingredient,
            parent,
            false
        )
        return IngredientViewHolder(containerView = view)
    }

    override fun onBindViewHolder(holder: IngredientViewHolder, position: Int) {
        val ingredientOnPosition = ingredientsToAdopt.get(index = position)
        holder.bind(ingredient = ingredientOnPosition)
    }

    inner class IngredientViewHolder(
        val containerView: View
    ) : RecyclerView.ViewHolder(containerView) {

        fun bind(ingredient: CreateOrEditRecipe.Ingredient) {
            containerView.addIngredient.setText(ingredient.product.name)
            containerView.ingredientCount.setText(ingredient.countInRecipe.toString())

            containerView.plus.setOnClickListener {
                presenter.onIngredientPlusClick(ingredient)
            }

            containerView.minus.setOnClickListener {
                presenter.onIngredientMinusClick(ingredient)
            }

            containerView.ingredientCount.setOnFocusChangeListener { v, hasFocus ->
                if (!hasFocus) {
                    val count = containerView.ingredientCount.text.toString().toDouble()
                    presenter.onSetIngredientCount(ingredient, count)
                }
            }
        }
    }
}